### PR TITLE
Issue 9602 use salt lengths in parity with real kms

### PR DESCRIPTION
**Pull request by georgecodes**
Modified KMS to use specific salt lengths to PSS signatures in line with AWS KMS

## Motivation

I discovered a disparity between AWS KMS and localstack KMS during signing of some mtls 
messages and raised this issue

https://github.com/localstack/localstack/issues/9602

When constructing the parameters for a PSS signature, an appropriate salt length is selected, rather than allowing a default which proves to be non-deterministic.

## Testing

Signing a message with a probabilistic scheme, and then verifying the signature offline with, for example, OpenSSL. The verification will fail. Contrast with AWS KMS, where it will not.

### CODE REVIEW

**Comment by georgecodes** [comment]
Unable to add a semver label as none are available to me. It's a patch though.

**Comment by sannya-singal** [comment]
Hi @georgecodes, thanks for contributing to localstack 🚀. Kindly add AWS validated tests to the changes similar to the ones present [here](https://github.com/localstack/localstack/blob/master/tests/aws/services/kms/test_kms.py).

**Comment by georgecodes** [comment]
Yep no problemOn 4 Jan 2024, at 07:38, Sannya Singal ***@***.***> wrote:﻿
Hi @georgecodes, thanks for contributing to localstack 🚀. Kindly add AWS validated tests to the changes similar to the ones present here.

File: tests/aws/services/kms/test_kms.py
Comment (sannya-singal) [inline tests/aws/services/kms/test_kms.py:93]: kindly remove the print statements being used.

File: tests/aws/services/kms/test_kms.py
Comment (sannya-singal) [inline tests/aws/services/kms/test_kms.py:92]: we need to add the logic in `kms/models.py`

File: tests/aws/services/kms/test_kms.py
Comment (sannya-singal) [inline tests/aws/services/kms/test_kms.py:483]: need to add snapshot validation for these changes.

**Review by sannya-singal — CHANGES_REQUESTED** [review]

File: tests/aws/services/kms/test_kms.py
Comment (sannya-singal) [reply tests/aws/services/kms/test_kms.py:92]: we usually write such logics in `models.py` since this piece of code can be used by other backend parts we shouldn't write it in test files. 

Eg: When we `decrypt` a kms key, we add the logic part of [here](https://github.com/localstack/localstack/blob/e393a7fdeeab4f178540d0be9daf60e438dd68af/localstack/services/kms/models.py#L273-L279) in the `kms/models.py`, and [test](https://github.com/localstack/localstack/blob/e393a7fdeeab4f178540d0be9daf60e438dd68af/tests/aws/services/kms/test_kms.py#L515-L517) just calls the handler which further uses the decrypt logic added in models.

**Comment by sannya-singal** [comment]
Hi @georgecodes! I have added some more context to the comments. Kindly refer to our [contribution guidelines](https://docs.localstack.cloud/contributing/) and let us know if you need some further assistance. 

Also please make sure that all the checks in the pipeline are green! 
Happy to help if you come across any further issues 🙂

File: tests/aws/services/kms/test_kms.py
Comment (FlorinAsavoaie) [reply tests/aws/services/kms/test_kms.py:483]: In fact I think the author of this PR is right in testing this way. Testing by doing the verification offline using the Public Key will clearly show why the current behaviour of LocalStack is NOT actually compatible with AWS, even though it claims so. RSAPSS signing in LocalStack is 100% not compatible with AWS KMS.

File: tests/aws/services/kms/test_kms.py
Comment (FlorinAsavoaie) [reply tests/aws/services/kms/test_kms.py:92]: This logic is definitely only required in testing. Why would testing-only code go into an application file?

**Comment by FlorinAsavoaie** [comment]
This is a significant bug that causes severe incompatibility with AWS KMS when using RSAPSS signing. Basically the code is 100% NOT compatible if you don't do the verification in KMS (which clearly cannot be done in lots of apps - as verification is usually done by other peers).

I am interested to improve this PR as needed until it is merged if the original author cannot continue to work on it.

**Comment by sannya-singal** [comment]
Hey @FlorinAsavoaie thanks for your interest in resolving this bug 🙌 🎉 Kindly go ahead and work on this fix: if you are planning to re-use a part of code by @georgecodes by cherry picking the commits, don't forget to give him the credits 😉 Thanks 🚀

**Comment by georgecodes** [comment]
> This is a significant bug that causes severe incompatibility with AWS KMS when using RSAPSS signing. Basically the code is 100% NOT compatible if you don't do the verification in KMS (which clearly cannot be done in lots of apps - as verification is usually done by other peers).
> 
> I am interested to improve this PR as needed until it is merged if the original author cannot continue to work on it.

Hi, I would love to spend some time on this but I am busy. I'm happy to revisit and address the comments later, but for now I have solved the problem by re-implementing KMS in pure Java, as I needed mocked KMS, and only KMS, for the specific work I was doing at the time I raised the issue.

But I'm also happy for you to pick this up if you have the time, naturally.

**Comment by FlorinAsavoaie** [comment]
@georgecodes would you mind trying to rebase the PR and removing that printed comment?

Other than that, the PR should be good to go probably. If not, I can definitely cherry-pick your code but it sounds faster and more accurate from a recognition perspective if you do this, doesn't look like more than a couple minutes thing.

File: tests/aws/services/kms/test_kms.py
Comment (sannya-singal) [reply tests/aws/services/kms/test_kms.py:92]: This should already be solved here: https://github.com/localstack/localstack/blob/f6c21e30cd112832e4b56cd6831df7e5b947c666/localstack/services/kms/models.py#L394 and https://github.com/localstack/localstack/blob/f6c21e30cd112832e4b56cd6831df7e5b947c666/localstack/services/kms/models.py#L416

File: tests/aws/services/kms/test_kms.py
Comment (sannya-singal) [reply tests/aws/services/kms/test_kms.py:483]: We can rather use the existing code as mentioned in the above comments.

**Comment by georgecodes** [comment]
@FlorinAsavoaie I will do so, yes

**Comment by sannya-singal** [comment]
Hey @georgecodes @FlorinAsavoaie 👋 Wanted to follow up on this PR, are you actively working on this? And if you would like some more inputs on further steps?

**Comment by FlorinAsavoaie** [comment]
This is not a priority for me at the moment, but the bug itself should probably be fixed as the behavior is significantly different comparing to AWS KMS.

**Comment by sannya-singal** [comment]
Thanks for letting me know @FlorinAsavoaie, this is already a part of our backlog. I will wait for @georgecodes response. Thanks 🙌

**Comment by sannya-singal** [comment]
Thank you for your contributions! I’m closing this PR for now, as no one is currently working on it. This issue shall be addressed in a future follow-up PR. However, if you’d like to continue working on it, please don’t hesitate to open another PR—I’d be happy to assist you with it.

**Pull request closed without merging**
